homeassistant/config_entries.py

```diff
@@ -759,13 +772,42 @@ async def async_reload(self, entry_id: str) -> bool:
 
         If an entry was not loaded, will just load.
         """
+        entry = self.async_get_entry(entry_id)
+
+        if entry is None:
+            raise UnknownEntry
+
         unload_result = await self.async_unload(entry_id)
 
-        if not unload_result:
+        if not unload_result or entry.disabled_by:
             return unload_result
 
         return await self.async_setup(entry_id)
 
+    async def async_set_disabled_by(
+        self, entry_id: str, disabled_by: Optional[str]
+    ) -> bool:
+        """Disable an entry.
+
+        If disabled_by is changed, the config entry will be reloaded.
+        """
+        entry = self.async_get_entry(entry_id)
+
+        if entry is None:
+            raise UnknownEntry
+
+        if entry.disabled_by == disabled_by:
+            return True
+
+        entry.disabled_by = disabled_by
+        self._async_schedule_save()
+
+        self.hass.bus.async_fire(
+            EVENT_CONFIG_ENTRY_DISABLED_BY_UPDATED, {"config_entry_id": entry_id}
```

Comment:
This event causes an error in combination with platform unload when the integration is reloaded.

When the entity registry updates the entity entries it will remove the entities since the entity entry is disabled. Since the entitieshave already been removed by platform unload, removing the entities again will trigger an error.

```
2021-02-26 23:53:55 ERROR (MainThread) [homeassistant] Error doing job: Task exception was never retrieved
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 638, in _async_registry_updated
    await self.async_remove()
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 545, in async_remove
    raise HomeAssistantError(
homeassistant.exceptions.HomeAssistantError: Entity switch.metered_wall_plug_switch async_remove called twice
```

https://github.com/home-assistant/core/issues/47171